mycpp/gc_mylib.h

```diff
@@ -41,14 +41,33 @@ Tuple2<Str*, Str*> split_once(Str* s, Str* delim);
 
 template <typename K, typename V>
 void dict_erase(Dict<K, V>* haystack, K needle) {
-  int pos = haystack->position_of_key(needle);
-  if (pos == -1) {
+  DCHECK(haystack->obj_header().heap_tag != HeapTag::Global);
+  int pos = haystack->hash_and_probe(needle);
+  if (pos == kNotFound || haystack->index_->items_[pos] < 0) {
+    // Nothing to do.
     return;
   }
-  haystack->entry_->items_[pos] = kDeletedEntry;
+
+  int offset = haystack->index_->items_[pos];
+  int last_offset = haystack->len_ - 1;
+  DCHECK(offset <= last_offset);
+
+  // To keep things compact, swap the target entry with the most recently
```

Comment:
Does this break any ordering guarantees?

- the last kv_index of `haystack_` is supposed to be the most recently inserted one?
- but then swapping it breaks this property?


Or maybe this just needs some comments about what the invariant is

Reply:
Yes, this does. Just like in CPython 3. I thought I documented this somewhere... I was under the impression we were OK breaking ordering slightly in the presence of deletions. Do you want to make this try harder to preserve ordering?

Reply:
Hmm that wasn't my understanding of Python 3

Where did you see that?

I thought it always preserved insertion order

Reply:
I thought you pointed that out somewhere on Zulip. There was some mailing list entry with some code snippets that did this. This compaction thing isn't strictly necessary, though. We can just remove it from "fast path" and handle compaction whenever we rehash

Reply:
Yeah let's handle it when we rehash, and also add a unit test for our insertion order guarantee

The tradeoff appears to be 

1. saving space on deletion, vs.
2. dict ordering

And I think matching CPython with (2) is more important

My memory is that we saw at least 2 different ways to maintain the guarantee in the face of deletions

But yeah I agree the simplest thing is to handle it on rehashing